extensions/amp-ad/0.1/amp-ad-ui.js

```diff
@@ -149,6 +168,85 @@ export class AmpAdUIHandler {
     return uiComponent;
   }
 
+  /**
+   * @return {boolean}
+   */
+  isStickyAd() {
+    return this.isStickyAd_;
+  }
+
+  /**
+   * Initialize sticky ad related features
+   */
+  maybeInitStickyAd() {
+    if (this.isStickyAd_) {
+      setStyles(this.element_, {
+        position: 'fixed',
+        bottom: '0',
+        right: '0',
+      });
+
+      this.element_.classList.add('i-amphtml-amp-ad-sticky-layout');
+    }
+  }
+
+  /**
+   * Scroll promise for sticky ad
+   * @return {Promise}
+   */
+  getScrollPromiseForStickyAd() {
+    if (this.isStickyAd_) {
+      return new Promise((resolve) => {
+        const unlisten = Services.viewportForDoc(
+          this.element_.getAmpDoc()
+        ).onScroll(() => {
+          resolve();
+          unlisten();
+        });
+      });
+    }
+    return Promise.resolve(null);
+  }
+
+  /**
+   * When a sticky ad is shown, the close button should be rendered at the same time.
+   */
+  onResizeSuccess() {
+    if (this.isStickyAd_ && !this.closeButtonRendered_) {
+      this.addCloseButton_();
+      this.closeButtonRendered_ = true;
+    }
+  }
+
+  /**
+   * The function that add a close button to sticky ad
+   */
+  addCloseButton_() {
+    const closeButton = createElementWithAttributes(
+      /** @type {!Document} */ (this.element_.ownerDocument),
+      'button',
+      dict({
+        'aria-label':
+          this.element_.getAttribute('data-close-button-aria-label') ||
+          'Close this ad',
+      })
+    );
+
+    this.unlisteners_.push(
+      listen(closeButton, 'click', () => {
+        Services.vsyncFor(this.baseInstance_.win).mutate(() => {
```

Comment:
should we also clean up the listener on click?

Reply:
Thanks for the reminder. I did forget to clean up. Though I feel cleaning up on destroying the frame is enough.

Reply:
ah didn't realize the iframe is getting destroyed right below.